### ISSUE TITLE
invalid_upcast_comparisons: false positive on implicit wrapping

### ISSUE THREAD

**Issue by toku-sa-n** [rust-lang/rust-clippy#7648]
Lint name: invalid_upcast_comparisons

I tried this [code](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=dc12224caa7a25817074e57f8216a4d0):

```rust
#![warn(clippy::invalid_upcast_comparisons)]
fn main() {
    let a = -1;
    let b = a as u64;

    println!("{}", (b as i32) < 0);
}
```

I expected to see this happen: Clippy says nothing.
Instead, this happened: 
```
warning: because of the numeric bounds on `b` prior to casting, this expression is always false
 --> src/main.rs:6:20
  |
6 |     println!("{}", (b as i32) < 0);
  |                    ^^^^^^^^^^^^^^
  |
note: the lint level is defined here
 --> src/main.rs:1:9
  |
1 | #![warn(clippy::invalid_upcast_comparisons)]
  |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#invalid_upcast_comparisons
```

### Meta

**Rust version (`rustc -Vv`):**
```
rustc 1.57.0-nightly (97032a6df 2021-09-08)
binary: rustc
commit-hash: 97032a6dfacdd3548e4bff98c90a6b3875a14077
commit-date: 2021-09-08
host: x86_64-unknown-linux-gnu
release: 1.57.0-nightly
LLVM version: 13.0.0
```

Clippy complained nothing if I change `u64` to `u32`.
Related: https://github.com/rust-lang/rust-clippy/issues/886
